File: python/day13.py
from utilities import read_lines
from typing import List, Tuple


def part_1(earliest_depart: int, buses: List[Tuple[int, int]]) -> int:
    wait = 10_000_00
    bus_id = None
    for _, bus in buses:
        if earliest_depart % bus == 0:
            bus_id = bus
            wait = 0
        else:
            depart = (earliest_depart // bus + 1) * bus
            if depart - earliest_depart < wait:
                wait = depart - earliest_depart
                bus_id = bus
    return wait * bus_id


def part_2(buses: List[Tuple[int, int]]) -> int:
    result = 0
    product = 1
    for _, bus in buses:
        product *= bus

    for index, bus in buses:
        b = (bus - index % bus) % bus
        n = product // bus
        x = 1
        xi = n
        while xi % bus != 1:
            xi += n
            x += 1

        result += b * n * x

    return result % product


if __name__ == '__main__':
    print('Day 13')
    data = read_lines('../inputs/day13.txt')
    earliest_depart = int(data[0])
    buses = data[1].split(',')
    buses = filter(lambda x: x[1] != 'x', enumerate(buses))
    buses = list(map(lambda x: (x[0], int(x[1])), buses))
    print('\tPart 1: {}'.format(part_1(earliest_depart, buses)))
    print('\tPart 2: {}'.format(part_2(buses)))

File: rust/src/day13.rs
use crate::utilities::read_lines;

fn part_1(earliest_depart: u64, buses: &Vec<(u64, u64)>) {
    let mut min_wait = 10_000_000;
    let mut bus_id = 0;
    for bus in buses {
        if earliest_depart % bus.0 == 0 {
            bus_id = bus.0;
            min_wait = 0;
            break;
        } else {
            let depart = (earliest_depart / bus.0 + 1) * bus.0;
            let wait = depart - earliest_depart;
            if wait < min_wait {
                min_wait = wait;
                bus_id = bus.0;
            }
        }
    }
    println!("\tPart 1: {}", min_wait * bus_id);
}

fn part_2(buses: &Vec<(u64, u64)>) {
    let product = buses.iter().fold(1, |acc, bus| acc * bus.0);
    let mut result = 0;

    for &(bus_id, wait) in buses {
        let b = (bus_id - wait % bus_id) % bus_id;
        let n = product / bus_id;
        let mut xi = n;
        let mut x = 1;
        while xi % bus_id != 1 {
            xi += n;
            x += 1;
        }
        result += b * n * x;
    }

    result %= product;

    println!("\tPart 2: {}", result);
}

pub fn main() {
    println!("Day 13");
    let data = read_lines("../inputs/day13.txt");
    let earliest_depart: u64 = data[0].parse().unwrap();
    let buses: Vec<(u64, u64)> = data[1]
        .split(',')
        .enumerate()
        .filter(|x| x.1 != "x")
        .map(|x| (x.1.parse::<u64>().unwrap(), x.0 as u64))
        .collect();

    part_1(earliest_depart, &buses);
    part_2(&buses);
}
